lib/btree.js

```diff
@@ -0,0 +1,240 @@
+  let [found, startIndex] = binarySearch(currNode, start);
+  if (found) startIndex++;
+  let finishIndex = finish === undefined ?
+    currNode.length - 1 :
+    binarySearch(currNode, finish)[1];
+  if (isLeaf(currNode)) {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      yield currNode[currIndex].data;
+    }
+  } else {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      const currElement = currNode[currIndex];
+      yield* inorderTraversal(start, finish, currElement.child);
+      yield currElement.data;
+    }
+    yield* inorderTraversal(start, finish, currNode[finishIndex].child);
+  }
+}
+
+const joinNodes = (parent, firstNodeIndex, secondNodeIndex) => {
+  const firstNode = parent[firstNodeIndex].child;
+  const secondNode = parent[secondNodeIndex].child;
+  const mid = parent.splice(firstNodeIndex, 1)[0];
+  mid.child = firstNode.pop().child;
+  secondNode.unshift(mid);
+  secondNode.splice(0, 0, ...firstNode);
+  return secondNode;
+};
+
+const growChild = (parent, childIndex) => {
+  const parentElement = parent[childIndex];
+  const node = parentElement.child;
+  const minDegree = node.length;
+  if (childIndex > 0) { // If we have left neighbor
+    const leftNeighbor = parent[childIndex - 1].child;
+    if (leftNeighbor.length > minDegree) {
+      const extractedElement =
+        leftNeighbor.splice(leftNeighbor.length - 2, 1)[0];
+      const insertedElement = new Element(
+        parent[childIndex - 1].value,
+        parent[childIndex - 1].data,
+        leftNeighbor[leftNeighbor.length - 1].child
+      );
+      leftNeighbor[leftNeighbor.length - 1].child = extractedElement.child;
+      extractedElement.child = leftNeighbor;
+      parent.splice(childIndex - 1, 1, extractedElement);
+      node.unshift(insertedElement);
+      return node;
+    }
+  } else if (childIndex < parent.length - 1) { // If we have right neighbor
+    const rightNeighbor = parent[childIndex + 1].child;
+    if (rightNeighbor.length > minDegree) {
+      const extractedElement = rightNeighbor.shift();
+      const insertedElement = new Element(
+        parentElement.value,
+        parentElement.data,
+        node[node.length - 1].child
+      );
+      node[node.length - 1].child = extractedElement.child;
+      extractedElement.child = node;
+      parent.splice(childIndex, 1, extractedElement);
+      node.splice(node.length - 1, 0, insertedElement);
+      return node;
+    } else return joinNodes(parent, childIndex, childIndex + 1);
+  }
+  return joinNodes(parent, childIndex - 1, childIndex);
+};
+
+const extractUpperLimit = (node, minDegree) => {
+  let currNode = node;
+  while (!(isLeaf(currNode))) {
+    let nextNode = currNode[0].child;
+    if (nextNode.length === minDegree) {
+      nextNode = growChild(currNode, 0);
+    }
+    currNode = nextNode;
+  }
+  return currNode.shift();
+};
+
+const extractLowerLimit = (node, minDegree) => {
```

Comment:
I think this and the function above can be merged via a single `i` argument (or bool argument and set i in the beginnig of the function) that will be 0 if Upper or currNode.length - 1 if Lower. And simple if check in the end.

Reply:
The index is always the same in the ``extractUpperLimit``, it is equal to 0, but in the ``extractLoverLimit`` it changes every iteration, so, every time when we use the index, it will be necessary to check which limit we need. Is it good?